dash_auth/group_protection.py

```diff
@@ -0,0 +1,198 @@
+    :return: None or list[str]:
+        * None if the user is not authenticated
+        * list[str] otherwise
+    """
+    if not has_request_context() or "user" not in session:
+        return None
+
+    user_groups = session.get("user", {}).get(groups_key, [])
+    # Handle cases where groups are ,- or ;-separated string,
+    # may depend on OIDC provider
+    if isinstance(user_groups, str) and groups_str_split is not None:
+        user_groups = re.split(groups_str_split, user_groups)
+    return user_groups
+
+
+def check_groups(
+    groups: Optional[List[str]] = None,
+    *,
+    groups_key: str = "groups",
+    groups_str_split: str = None,
+    check_type: CheckType = "one_of",
+) -> Optional[bool]:
+    """Check whether the current user is authenticated
+    and has the specified groups.
+
+    :param groups: List of groups to check for with check_type
+    :param groups_key: Groups key in the user data saved in the Flask session
+        e.g. session["user"] == {"email": "a.b@mail.com", "groups": ["admin"]}
+    :param groups_str_split: Used to split groups if provided as a string
+    :param check_type: Type of check to perform.
+        Either "one_of", "all_of" or "none_of"
+    :return: None or boolean:
+        * None if the user is not authenticated
+        * True if the user is authenticated and has the right permissions
+        * False if the user is authenticated but does not have
+          the right permissions
+    """
+    user_groups = list_groups(
+        groups_key=groups_key,
+        groups_str_split=groups_str_split,
+    )
+
+    if user_groups is None:
+        # User is not authenticated
+        return None
+
+    if groups is None:
+        return True
+
+    if check_type == "one_of":
+        return bool(set(user_groups).intersection(groups))
+    if check_type == "all_of":
+        return all(group in user_groups for group in groups)
+    if check_type == "none_of":
+        return not any(group in user_groups for group in groups)
+
+    raise ValueError(f"Invalid check_type: {check_type}")
+
+
+def protected(
+    unauthenticated_output: Optional[OutputVal] = None,
+    *,
+    missing_permissions_output: Optional[OutputVal] = None,
+    groups: Optional[List[str]] = None,
+    groups_key: str = "groups",
+    groups_str_split: str = None,
+    check_type: CheckType = "one_of",
+) -> Callable:
+    """Decorate a function or output to alter it depending on the state
+    of authentication and permissions.
+
+    :param unauthenticated_output: Output when the user is not authenticated.
+        Note: needs to be a function with no argument or
+        a collection of Dash components.
+    :param missing_permissions_output: Output when the user is authenticated
+        but does not have the right permissions.
+        It defaults to unauthenticated_output when not set.
+        Note: needs to be a function with no argument or
+        a collection of Dash components.
```

Comment:
Maybe I'm not getting the purpose of `protected`, as opposed to `protected_callback` - why would the fallback outputs here want to be Dash components? It looks like this could be used for any function, not just one that returns Dash components or props

Reply:
You are totally right it should be Any here.